packages/opal-client/opal_client/data/updater.py

```diff
@@ -334,197 +330,185 @@ async def stop(self):
-                            opal_client_config.SPLIT_ROOT_DATA
-                            and policy_store_path in ("/", "")
-                            and isinstance(policy_data, dict)
-                        ):
-                            await self._set_split_policy_data(
-                                store_transaction,
-                                url=url,
-                                save_method=entry.save_method,
-                                data=policy_data,
-                            )
-                        else:
-                            await self._set_policy_data(
-                                store_transaction,
-                                url=url,
-                                path=policy_store_path,
-                                save_method=entry.save_method,
-                                data=policy_data,
-                            )
-                        # No exception we we're able to save to the policy-store
-                        report.saved = True
-                        # save the report for the entry
-                        reports.append(report)
-                    except Exception:
-                        logger.exception("Failed to save data update to policy-store")
-                        # we failed to save to policy-store
-                        report.saved = False
-                        # save the report for the entry
-                        reports.append(report)
-                        # re-raise so the context manager will be aware of the failure
-                        raise
-                else:
-                    report = DataEntryReport(entry=entry, fetched=False, saved=False)
-                    # save the report for the entry
-                    reports.append(report)
-        # should we send a report to defined callbackers?
+    async def _send_reports(self, reports: list[DataEntryReport], update: DataUpdate):
         if self._should_send_reports:
             # spin off reporting (no need to wait on it)
             whole_report = DataUpdateReport(update_id=update.id, reports=reports)
             extra_callbacks = self._callbacks_register.normalize_callbacks(
                 update.callback.callbacks
             )
-            self._tasks.add_task(
+            self._tasks.create_task(
                 self._callbacks_reporter.report_update_results(
                     whole_report, extra_callbacks
                 )
             )
 
+    async def _fetch_and_save_data(
+        self,
+        entry: DataSourceEntry,
+        store_transaction: PolicyStoreTransactionContextManager,
+    ) -> DataEntryReport:
+        """Fetches data from the given entry and saves it to the policy store.
+
+        :param entry: The data source entry
+        :param store_transaction: The policy store transaction
+        :return: The data entry report
+        """
+        try:
+            result = await self._data_fetcher.handle_url(
```

Comment:
You killed the performance of this thing - you made all data fetching serial; 
There no longer is a separate task for handling fetching, and the update task is singular.

Reply:
This is the biggest issue here

Reply:
No, data updates from the root are done concurrently - with lock on each destination. Non-conflicting destinations (the common ones) will be done in parallel

See line 170 - 
```python
self._tasks.create_task(self._update_policy_data(update))
```
We await the lock for this dst, then immediately start fetching the data, then set it to the policy store. This flow is done concurrently between all data updates

https://github.com/permitio/opal/pull/745/files/06a3324f1adb22794cbdeb0f15366d9a988e0294#diff-16614de9db264ea3ec0d36343b453cc9881371fe95ff4a862267eaaf24c563c2R170

Reply:
You are right; I'm sorry I missed this (In my defence the diff here is extremely hard to read )